ocaml/runtime/shared_heap.c

```diff
@@ -958,8 +1018,417 @@ static void compact_update_ephe_list(volatile value *ephe_p)
   }
 }
 
+int compact_count_pools(pool* start_pool) {
+  int count = 0;
+  while( start_pool ) {
+    count++;
+    start_pool = start_pool->next;
+  }
+
+  return count;
+}
+
+/* sort first by chunk_size and then by chunk */
+int compact_compare_pools(const void* v1, const void* v2) {
+  pool* p1 = *(pool**)v1;
+  pool* p2 = *(pool**)v2;
+
+  if (p1->chunk_size < p2->chunk_size) {
+    return 1;
+  } else if (p1->chunk_size > p2->chunk_size) {
+    return -1;
+  } else {
+    if (p1->chunk > p2->chunk) {
+      return -1;
+    } else if (p1->chunk < p2->chunk) {
+      return 1;
+    } else {
+      /* otherwise we want the pools in address order (which helps freeing them later) */
+      return p2 > p1;
```

Comment:
Good spot. Fixed.

I'm not sure I understand how sorting in to increased address order would simplify `compact_release_freelist`. I think we need to resort at that point anyway because the previous step is parallel and we can't assume the resulting freelist is in order?

Reply:
The iteration over the free pools, identifying ones which form a complete chunk so that we can then unmap the chunk, might be simpler if the pools array were in _increasing_ rather than _decreasing_ address order. But it's not a big deal.